# Data Manipulation
## Assignment: LEGO Project
### Name: Richard Lee

## Overview

In this assignment, you will load the various tables from the [Rebrickable](https://rebrickable.com/) website and perform analysis on the data.  Rebrickable is a fantastic website that is used to show you which LEGO sets you can build from the sets and parts you already own.  They will even show you which parts you are missing to complete the new set and suggest sets that you could purchase to get those parts.  

## Data

The data has been downloaded directly from their website and includes files for the tables shown in the schema above.  Your first task will be to go through the data and review the schema to understand how the tables all fit together.  There is no data dictionary that I am aware of, but you should be able to understand this data by reviewing the files and schema and visiting their [Download](https://rebrickable.com/downloads/) page and [API](https://rebrickable.com/api/) documentation. 

## Assignment

In this assignment, you will create a number of functions to complete certain tasks.

In addition, we would strongly suggest that you merge data sets as appropriate for the given task.  For example, you might merge the `sets` and `themes` DataFrames to answer one question, and then merge `sets`, `inventories`, `inventory_minifigs`, and `minifigs` to answer another.  This will make your code much easier to write and will be key in working with data across multiple tables.  In fact, I was able to create each function with only one line of code because of the merging of certain data sets.  Also note that you may not use all the tables in this assignment.  This is also an important skill for a Data Scientist to know when to use data and what data is unnecessary.


In [1]:
# common imports
import pandas as pd
import numpy as np

# Do not change this option; This allows the CodeGrade auto grading to function correctly
pd.set_option('display.max_columns', None)

## Import Data

Load the various Lego datasets below calling them the same as the name of the table as seen in the schema.  Note that you may or may not have to use all the data sets.  One of the things that all Data Scientists and Analysts have to decide when working on a project is what data is useful for a task and what data is not.  Remember to look at the schema above to see how all the data sets are connected.

In [2]:
# import Lego data sets here
# make sure that files are in the same folder as your notebook
inventories = pd.read_csv('inventories.csv')
inventory_sets = pd.read_csv('inventory_sets.csv')
sets = pd.read_csv('sets.csv')
themes = pd.read_csv('themes.csv')
inventory_minifigs = pd.read_csv('inventory_minifigs.csv')
minifigs = pd.read_csv('minifigs.csv')
inventory_parts = pd.read_csv('inventory_parts.csv')
part_categories = pd.read_csv('part_categories.csv') 
parts = pd.read_csv('parts.csv')
colors = pd.read_csv('colors.csv')
part_relationships = pd.read_csv('part_relationships.csv')
elements = pd.read_csv('elements.csv')

## Exploration

Take time to look at all of the data sets using `.head()`, `.info()`, `.describe()`, etc to familiarize yourself with the data sets.  As you look at them, notice how the various tables connect together.

In [3]:
def data(x):
    print(f'head: {x.head()}')
    print(f'info: {x.info()}')
    print(f'describe: {x.describe()}')

In [4]:
data(inventories)

head:    id  version set_num
0   1        1  7922-1
1   3        1  3931-1
2   4        1  6942-1
3  15        1  5158-1
4  16        1   903-1
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31587 entries, 0 to 31586
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   id       31587 non-null  int64 
 1   version  31587 non-null  int64 
 2   set_num  31587 non-null  object
dtypes: int64(2), object(1)
memory usage: 740.4+ KB
info: None
describe:                   id       version
count   31587.000000  31587.000000
mean    43538.531389      1.077405
std     32458.542329      0.515993
min         1.000000      1.000000
25%     12193.500000      1.000000
50%     51076.000000      1.000000
75%     62254.500000      1.000000
max    122472.000000     11.000000


In [5]:
data(inventory_sets)

head:    inventory_id  set_num  quantity
0            35  75911-1         1
1            35  75912-1         1
2            39  75048-1         1
3            39  75053-1         1
4            50   4515-1         1
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3413 entries, 0 to 3412
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   inventory_id  3413 non-null   int64 
 1   set_num       3413 non-null   object
 2   quantity      3413 non-null   int64 
dtypes: int64(2), object(1)
memory usage: 80.1+ KB
info: None
describe:         inventory_id     quantity
count    3413.000000  3413.000000
mean    25839.150601     1.857017
std     31384.971888     6.123060
min        35.000000     1.000000
25%      6117.000000     1.000000
50%     12465.000000     1.000000
75%     27003.000000     1.000000
max    122081.000000    60.000000


In [6]:
data(sets)

head:   set_num                        name  year  theme_id  num_parts
0   001-1                       Gears  1965         1         43
1  0011-2           Town Mini-Figures  1979        67         12
2  0011-3  Castle 2 for 1 Bonus Offer  1987       199          0
3  0012-1          Space Mini-Figures  1979       143         12
4  0013-1          Space Mini-Figures  1979       143         12
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18576 entries, 0 to 18575
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   set_num    18576 non-null  object
 1   name       18576 non-null  object
 2   year       18576 non-null  int64 
 3   theme_id   18576 non-null  int64 
 4   num_parts  18576 non-null  int64 
dtypes: int64(3), object(2)
memory usage: 725.8+ KB
info: None
describe:                year      theme_id     num_parts
count  18576.000000  18576.000000  18576.000000
mean    2006.200581    405.924365    160.666344
std   

In [7]:
data(themes)

head:    id            name  parent_id
0   1         Technic        NaN
1   3     Competition        1.0
2   4  Expert Builder        1.0
3  16      RoboRiders        1.0
4  17  Speed Slammers        1.0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 444 entries, 0 to 443
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   id         444 non-null    int64  
 1   name       444 non-null    object 
 2   parent_id  306 non-null    float64
dtypes: float64(1), int64(1), object(1)
memory usage: 10.5+ KB
info: None
describe:                id   parent_id
count  444.000000  306.000000
mean   410.659910  346.516340
std    208.165803  198.728717
min      1.000000    1.000000
25%    229.750000  186.000000
50%    438.500000  409.500000
75%    597.250000  507.000000
max    719.000000  697.000000


In [8]:
data(inventory_minifigs)

head:    inventory_id     fig_num  quantity
0             3  fig-001549         1
1             4  fig-000764         1
2            19  fig-000555         1
3            25  fig-000574         1
4            26  fig-000842         1
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18144 entries, 0 to 18143
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   inventory_id  18144 non-null  int64 
 1   fig_num       18144 non-null  object
 2   quantity      18144 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 425.4+ KB
info: None
describe:         inventory_id      quantity
count   18144.000000  18144.000000
mean    26470.198302      1.066138
std     30578.235290      0.831258
min         3.000000      1.000000
25%      6901.000000      1.000000
50%     13832.000000      1.000000
75%     32633.000000      1.000000
max    122344.000000    100.000000


In [9]:
data(minifigs)

head:       fig_num                             name  num_parts
0  fig-000001               Toy Store Employee          4
1  fig-000002                     Customer Kid          4
2  fig-000003            Assassin Droid, White          8
3  fig-000004                     Basic Figure          4
4  fig-000005  Captain America with Short Legs          3
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11827 entries, 0 to 11826
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   fig_num    11827 non-null  object
 1   name       11827 non-null  object
 2   num_parts  11827 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 277.3+ KB
info: None
describe:           num_parts
count  11827.000000
mean       4.847637
std        3.562961
min        0.000000
25%        4.000000
50%        4.000000
75%        5.000000
max      148.000000


In [10]:
data(inventory_parts)

head:    inventory_id        part_num  color_id  quantity is_spare
0             1        48379c01        72         1        f
1             1           48395         7         1        f
2             1  stickerupn0077      9999         1        f
3             1         upn0342         0         1        f
4             1         upn0350        25         1        f
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 989529 entries, 0 to 989528
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   inventory_id  989529 non-null  int64 
 1   part_num      989529 non-null  object
 2   color_id      989529 non-null  int64 
 3   quantity      989529 non-null  int64 
 4   is_spare      989529 non-null  object
dtypes: int64(3), object(2)
memory usage: 37.7+ MB
info: None
describe:         inventory_id       color_id       quantity
count  989529.000000  989529.000000  989529.000000
mean    30715.103387     124.593946       3

In [11]:
data(part_categories)

head:    id                     name
0   1               Baseplates
1   3            Bricks Sloped
2   4  Duplo, Quatro and Primo
3   5           Bricks Special
4   6            Bricks Wedged
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66 entries, 0 to 65
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   id      66 non-null     int64 
 1   name    66 non-null     object
dtypes: int64(1), object(1)
memory usage: 1.2+ KB
info: None
describe:               id
count  66.000000
mean   35.363636
std    19.411049
min     1.000000
25%    19.250000
50%    35.500000
75%    51.750000
max    68.000000


In [12]:
data(parts)

head:   part_num                                        name  part_cat_id  \
0   003381                 Sticker Sheet for Set 663-1           58   
1   003383         Sticker Sheet for Sets 618-1, 628-2           58   
2   003402  Sticker Sheet for Sets 310-3, 311-1, 312-3           58   
3   003429                Sticker Sheet for Set 1550-1           58   
4   003432  Sticker Sheet for Sets 357-1, 355-1, 940-1           58   

  part_material  
0       Plastic  
1       Plastic  
2       Plastic  
3       Plastic  
4       Plastic  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44781 entries, 0 to 44780
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   part_num       44781 non-null  object
 1   name           44781 non-null  object
 2   part_cat_id    44781 non-null  int64 
 3   part_material  44781 non-null  object
dtypes: int64(1), object(3)
memory usage: 1.4+ MB
info: None
describe:         part_cat_id
co

In [13]:
data(colors)

head:    id            name     rgb is_trans
0  -1       [Unknown]  0033B2        f
1   0           Black  05131D        f
2   1            Blue  0055BF        f
3   2           Green  237841        f
4   3  Dark Turquoise  008F9B        f
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 190 entries, 0 to 189
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   id        190 non-null    int64 
 1   name      190 non-null    object
 2   rgb       190 non-null    object
 3   is_trans  190 non-null    object
dtypes: int64(1), object(3)
memory usage: 6.1+ KB
info: None
describe:                 id
count   190.000000
mean    479.394737
std     820.735479
min      -1.000000
25%      61.250000
50%     182.500000
75%    1015.750000
max    9999.000000


In [14]:
data(part_relationships)

head:   rel_type child_part_num parent_part_num
0        P    87079pr9974           87079
1        P     3960pr9971            3960
2        R    98653pr0003     98086pr0003
3        R    98653pr0003    98088pat0003
4        R    98653pr0003    98089pat0003
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24508 entries, 0 to 24507
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   rel_type         24508 non-null  object
 1   child_part_num   24508 non-null  object
 2   parent_part_num  24508 non-null  object
dtypes: object(3)
memory usage: 574.5+ KB
info: None
describe:        rel_type child_part_num parent_part_num
count     24508          24508           24508
unique        6          22339            3858
top           P           3814           3626c
freq      19745             23            2353


In [15]:
data(elements)

head:   element_id  part_num  color_id
0    6300211  67906c01        14
1    4566309      2564         0
2    4275423     53657      1004
3    6194308     92926        71
4    6229123     26561         4
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62684 entries, 0 to 62683
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   element_id  62684 non-null  object
 1   part_num    62684 non-null  object
 2   color_id    62684 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 1.4+ MB
info: None
describe:            color_id
count  62684.000000
mean     608.309983
std     2210.989312
min       -1.000000
25%       14.000000
50%       30.000000
75%      148.000000
max     9999.000000


## Assignment

Answer all the questions below using the available data.  There are multiple ways to get some of the answers.  I would suggest that you attempt to merge some data sets as needed to assist with getting answers.  This will make your tasks a lot easier and your code will be cleaner.


**Q1:** Create a function called `data_count` that takes as input the data set (for example: `sets`) and returns the total number of rows in the data (as an integer).  Nothing else should be returned.

In [16]:
def data_count(y):
    return len(y)
    

**Code Check:**  If you call your `data_count` function using the `sets` DataFrame, you should see that there are 18,576 sets represented in the data.

In [17]:
data_count(sets)

18576

**Q2:** What is the average number of parts for a Lego set (rounded to the nearest integer)?  Save this as `Q2`.

In [18]:
data(sets)

head:   set_num                        name  year  theme_id  num_parts
0   001-1                       Gears  1965         1         43
1  0011-2           Town Mini-Figures  1979        67         12
2  0011-3  Castle 2 for 1 Bonus Offer  1987       199          0
3  0012-1          Space Mini-Figures  1979       143         12
4  0013-1          Space Mini-Figures  1979       143         12
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18576 entries, 0 to 18575
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   set_num    18576 non-null  object
 1   name       18576 non-null  object
 2   year       18576 non-null  int64 
 3   theme_id   18576 non-null  int64 
 4   num_parts  18576 non-null  int64 
dtypes: int64(3), object(2)
memory usage: 725.8+ KB
info: None
describe:                year      theme_id     num_parts
count  18576.000000  18576.000000  18576.000000
mean    2006.200581    405.924365    160.666344
std   

In [19]:
Q2 = int(round(sets['num_parts'].mean(),0))
Q2

161

**Q3:** What is the median number of parts per set (output as an integer)?  Save this as `Q3`.

In [20]:
Q3 = int(round(sets['num_parts'].median(),0))
Q3

37

You should see that the median number of parts is much lower than the average number of parts per set.  This tells us that there are some Lego sets with a very high number of parts that brings the average up.

**Q4:** What is the largest number of parts that a set has? Save this as `Q4`.

In [21]:
Q4 = sets['num_parts'].max()
Q4

11695

**Q5:** Given the set with the largest number of parts calculated above, select its row from the `sets` DataFrame.  Save this as `Q5`.  Do not reset the index.

In [22]:
Q5 = sets[sets['num_parts'] == Q4]
Q5

set_num       name  year  theme_id  num_parts
3867  31203-1  World Map  2021       709      11695

**Q6:** Create a function called `select_set_row` where you pass it the `set_num` (as a string) and the function returns the respective row of the DataFrame from `sets`. Do not reset the index.

In [23]:
sets

set_num                             name  year  theme_id  num_parts
0            001-1                            Gears  1965         1         43
1           0011-2                Town Mini-Figures  1979        67         12
2           0011-3       Castle 2 for 1 Bonus Offer  1987       199          0
3           0012-1               Space Mini-Figures  1979       143         12
4           0013-1               Space Mini-Figures  1979       143         12
...            ...                              ...   ...       ...        ...
18571      XWING-1              Mini X-Wing Fighter  2019       158         60
18572      XWING-2                X-Wing Trench Run  2019       158         52
18573  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158        413
18574   YTERRIER-1                Yorkshire Terrier  2018       598          0
18575     ZX8000-1             ZX 8000 LEGO Sneaker  2020       501          0

[18576 rows x 5 columns]

In [24]:
def select_set_row(a):
    return sets[sets['set_num'] == a]

In [25]:
select_set_row('XWING-1')
    

set_num                 name  year  theme_id  num_parts
18571  XWING-1  Mini X-Wing Fighter  2019       158         60

**Q7:** Create a function called `select_set_numparts` that takes as input a specific number of parts (as an integer) and selects all the rows in the `sets` DataFrame that contain those number of parts.  Return the DataFrame sorted by `year` and then by `set_num`.  Do not reset the index.

In [26]:
def select_set_numparts(b):
    return sets[sets['num_parts']== b].sort_values(['year', 'set_num'])


**Code Check:** Call the `select_set_numparts()` function to select all sets with `0` number of parts.  There should be 3,146 rows returned.

In [27]:
select_set_numparts(0)

set_num                                               name  \
11061          700.1-2                     Extra-Large Gift Set (Mursten)   
17563         b55dk-01               Lego Mursten - System i Leg Byggebog   
17564         b56de-01                                    System im Spiel   
11072          700.2-2                                       Gift Package   
11078         700.3A-2                                       Gift Package   
...                ...                                                ...   
15619         854195-1                                    Ernie Key Chain   
15620         854197-1                               Demogorgon Key Chain   
17170  9780744054583-1  DC Comics Super Heroes: Character Encyclopedia...   
17193  9780794447557-1                               City: Color The City   
17200  9780794448172-1           DC Comics Super Heroes: Ready For Action   

       year  theme_id  num_parts  
11061  1953       366          0  
17563  1955       497          0  
17564  1956       497          0  
11072  1957       366          0  
11078  1957       372          0  
...     ...       ...        ...  
15619  2022       503          0  
15620  2022       503          0  
17170  2022       497          0  
17193  2022       497          0  
17200  2022       497          0  

[3146 rows x 5 columns]

**Q8:** 
- The oldest set listed was made in what year?  Save the year as `Q8A` (as an integer).  
- The newest set listed was made in what year? Save the year as `Q8B` (as an integer).

In [28]:
Q8A = sets['year'].min()
Q8A

1949

In [29]:
Q8B = sets['year'].max()
Q8B

2022

**Q9:** 
- Create a function called `select_set_year` that takes as input a year (as an integer) and returns the rows from the `sets` DataFrame that matches that year.  Sort the DataFrame by `set_num`.  Do not reset the index.

In [30]:
def select_set_year(c):
    return sets[sets['year']== c].sort_values('set_num')

In [31]:
select_set_year(1949)

set_num                               name  year  theme_id  num_parts
11057  700.1-1         Extra-Large Gift Set (ABB)  1949       365        142
11070  700.2-1               Large Gift Set (ABB)  1949       365        178
11075  700.3-1              Medium Gift Set (ABB)  1949       365        142
11080  700.A-1              Small Brick Set (ABB)  1949       371         24
11081  700.B-1  Small Doors and Windows Set (ABB)  1949       371         12

**Q10:** 
- Create a function called `theme_by_year` that takes as input a year (as an integer) and shows the theme ids and theme names (listed in order by theme id) that were in sets that year.  
- The column names must be `id` and `name_themes` (to differentiate between the name of a theme and the name of a set) in that order.  
- The index should be reset and go from `0` to `n-1`.  
- Each theme should only be listed once even if it appeared in more than one set from that year -- duplicate themes should be based on theme id and not name since there are some themes with the same name but with a different id.
- Hint:  It will help if you were to think about merging appropriate DataFrames to help you get this answer.

In [32]:
themes.head()

id            name  parent_id
0   1         Technic        NaN
1   3     Competition        1.0
2   4  Expert Builder        1.0
3  16      RoboRiders        1.0
4  17  Speed Slammers        1.0

In [33]:
themesandsets = pd.merge(sets,themes, left_on = 'theme_id', right_on='id')
themesandsets.rename(columns={'name_y':'name_themes'}, inplace=True)

In [34]:

def theme_by_year(year):
    return themesandsets[themesandsets['year'] == year].drop_duplicates('theme_id')[['id', 
                                                                                     'name_themes']].sort_values('id').reset_index(drop=True)
    

    

In [35]:
themesandsets.loc[themesandsets['year'] == 1960]

set_num                       name_x  year  theme_id  num_parts   id  \
875     238-1  Lego System Idea Book no. 1  1960       497          0  497   
2348    237-1                Number Bricks  1960       371         50  371   
17578  700K-1        Kindergarten LEGO Set  1960       513        476  513   

        name_themes  parent_id  
875           Books        NaN  
2348   Supplemental      365.0  
17578       Classic      507.0

**Code Check:** Call the `theme_by_year()` function on the year `1960`.  Your output should look as follows:

| | id | name_themes|
|-|---|---|
|0|371|Supplemental|
|1|497|Books|
|2|513|Classic|

**Optional Practice:** Call the `theme_by_year()` function using your birth year to see the Lego themes that were popular when you were born.

In [36]:
theme_by_year(1960)

id   name_themes
0  371  Supplemental
1  497         Books
2  513       Classic

**Q11:** 
- Create a function called `theme_by_name` that takes as input a theme name and outputs all the sets (in `set_num` order) that had that theme.  
- The DataFrame should include the following columns in this order and with these exact names: `set_num`,`name_sets`,`year`,`num_parts`, `theme_id`, and `name_themes`.  
- The index should go from `0` to `n-1`.

In [37]:
themesandsets.rename(columns={'name_x':'name_sets'}, inplace=True)
def theme_by_name(e):
    return themesandsets[themesandsets['name_themes']== e][['set_num','name_sets', 'year','num_parts', 'theme_id', 'name_themes']].sort_values('set_num').reset_index(drop=True)
    

**Code Check:** Call the `theme_by_name()` function using the theme name `Gear`.  You should return 1,904 sets that match that theme name.

In [38]:
theme_by_name('Gear')

set_num                                          name_sets  \
0              11138-1                            Bionicle Barraki Wallet   
1       119645111436-1                Classic Space Astronaut Plush (Red)   
2       119645111504-1               Classic Space Astronaut Plush (Blue)   
3       119645111740-1              Classic Space Astronaut Plush (White)   
4       119645111818-1               Classic Space Astronaut Plush (Pink)   
...                ...                                                ...   
1899           WATCH-3                   Watch System Voyager (UFO Space)   
1900         WIIGAME-1                              Bionicle Heroes - Wii   
1901         WIIGAME-2  Indiana Jones: The Original Adventures Video G...   
1902  WOODENELEPHANT-1  Original Wooden Elephant Pull-Along Toy and Print   
1903          ZX8000-1                               ZX 8000 LEGO Sneaker   

      year  num_parts  theme_id name_themes  
0     2008          0       501        Gear  
1     2021          0       501        Gear  
2     2021          0       501        Gear  
3     2021          0       501        Gear  
4     2021          0       501        Gear  
...    ...        ...       ...         ...  
1899  1997         29       501        Gear  
1900  2007          0       501        Gear  
1901  2008          0       501        Gear  
1902  2021          0       501        Gear  
1903  2020          0       501        Gear  

[1904 rows x 6 columns]

**Q12:** Create a function called `theme_by_setnum` that takes as input the set_num (as a string) and returns the respective theme name as a string.

In [39]:
def theme_by_setnum(f):
    return themesandsets.loc[themesandsets['set_num'] == f, 'name_themes'].reset_index(drop=True).loc[0]
    

**Code Check:** Visit the [Sets page](https://rebrickable.com/sets/) from the Rebrickable website and find a set that you like.  Call your `theme_by_setnum()` function on that set and make sure the correct theme name is output.

In [40]:
theme_by_setnum('WATCH-3')

'Gear'

**Millennium Falcon Prep Exercise:** Find all the sets named 'Millennium Falcon'. Of these sets, find the set with the largest number of parts.  You will use this information for later questions.

In [41]:
themesandsets[themesandsets['name_sets']== 'Millennium Falcon']

set_num          name_sets  year  theme_id  num_parts   id  \
3177       10179-1  Millennium Falcon  2007       171       5197  171   
3199       75192-1  Millennium Falcon  2017       171       7541  171   
11849       4488-1  Millennium Falcon  2003       158         87  158   
11995       7190-1  Millennium Falcon  2000       158        671  158   
12042      75030-1  Millennium Falcon  2014       158         94  158   
12099      75105-1  Millennium Falcon  2015       158       1330  158   
12237      75257-1  Millennium Falcon  2019       158       1351  158   
12361       7965-1  Millennium Falcon  2011       158       1254  158   
12398     911607-1  Millennium Falcon  2016       158         42  158   
12440     911949-1  Millennium Falcon  2019       158         32  158   
12542  TRUFALCON-1  Millennium Falcon  2016       158         44  158   
12545    TRUSWMF-2  Millennium Falcon  2018       158         24  158   

                     name_themes  parent_id  
3177   Ultimate Collector Series      158.0  
3199   Ultimate Collector Series      158.0  
11849                  Star Wars        NaN  
11995                  Star Wars        NaN  
12042                  Star Wars        NaN  
12099                  Star Wars        NaN  
12237                  Star Wars        NaN  
12361                  Star Wars        NaN  
12398                  Star Wars        NaN  
12440                  Star Wars        NaN  
12542                  Star Wars        NaN  
12545                  Star Wars        NaN

**Q13:** 
- How many minifigs are contained in the set from the `Millennium Falcon Prep Exercise`?  Save this as `Q13` (as an integer).  
- Note that there are sometimes multiple versions of a set as shown in the `inventories` table.  This will not matter for this question as there is only one version for the respective set but this should be noted for future questions.
- Remember to show your work or your answer will manually get marked as incorrect.

In [42]:
themesandsets

set_num                           name_sets  year  theme_id  \
0            001-1                               Gears  1965         1   
1            002-1      4.5V Samsonite Gears Motor Set  1965         1   
2           1030-1      TECHNIC I: Simple Machines Set  1985         1   
3           1038-1                  ERBIE the Robo-Car  1985         1   
4           1039-1                Manual Control Set 1  1986         1   
...            ...                                 ...   ...       ...   
18571  M20-2566-10  Modulex Box - 10x Window 1 x 6 x 6  1963       716   
18572  M20-2575-10  Modulex Box - 10x Window 1 x 7 x 5  1963       716   
18573  M20-2576-10  Modulex Box - 10x Window 1 x 7 x 6  1963       716   
18574  M20-2586-10  Modulex Box - 10x Window 1 x 8 x 6  1963       716   
18575  WEETABIX1-1                     Weetabix Castle  1970       414   

       num_parts   id name_themes  parent_id  
0             43    1     Technic        NaN  
1              3    1     Technic        NaN  
2            211    1     Technic        NaN  
3            120    1     Technic        NaN  
4             39    1     Technic        NaN  
...          ...  ...         ...        ...  
18571         10  716     Modulex        NaN  
18572         10  716     Modulex        NaN  
18573         10  716     Modulex        NaN  
18574         10  716     Modulex        NaN  
18575        471  414      Castle      411.0  

[18576 rows x 8 columns]

In [43]:
themesandsets

set_num                           name_sets  year  theme_id  \
0            001-1                               Gears  1965         1   
1            002-1      4.5V Samsonite Gears Motor Set  1965         1   
2           1030-1      TECHNIC I: Simple Machines Set  1985         1   
3           1038-1                  ERBIE the Robo-Car  1985         1   
4           1039-1                Manual Control Set 1  1986         1   
...            ...                                 ...   ...       ...   
18571  M20-2566-10  Modulex Box - 10x Window 1 x 6 x 6  1963       716   
18572  M20-2575-10  Modulex Box - 10x Window 1 x 7 x 5  1963       716   
18573  M20-2576-10  Modulex Box - 10x Window 1 x 7 x 6  1963       716   
18574  M20-2586-10  Modulex Box - 10x Window 1 x 8 x 6  1963       716   
18575  WEETABIX1-1                     Weetabix Castle  1970       414   

       num_parts   id name_themes  parent_id  
0             43    1     Technic        NaN  
1              3    1     Technic        NaN  
2            211    1     Technic        NaN  
3            120    1     Technic        NaN  
4             39    1     Technic        NaN  
...          ...  ...         ...        ...  
18571         10  716     Modulex        NaN  
18572         10  716     Modulex        NaN  
18573         10  716     Modulex        NaN  
18574         10  716     Modulex        NaN  
18575        471  414      Castle      411.0  

[18576 rows x 8 columns]

In [44]:
inventories.head()

id  version set_num
0   1        1  7922-1
1   3        1  3931-1
2   4        1  6942-1
3  15        1  5158-1
4  16        1   903-1

In [45]:

mini1 = themesandsets.merge(inventories, on='set_num')
mini1

set_num                           name_sets  year  theme_id  \
0            001-1                               Gears  1965         1   
1            002-1      4.5V Samsonite Gears Motor Set  1965         1   
2           1030-1      TECHNIC I: Simple Machines Set  1985         1   
3           1038-1                  ERBIE the Robo-Car  1985         1   
4           1039-1                Manual Control Set 1  1986         1   
...            ...                                 ...   ...       ...   
19755  M20-2576-10  Modulex Box - 10x Window 1 x 7 x 6  1963       716   
19756  M20-2576-10  Modulex Box - 10x Window 1 x 7 x 6  1963       716   
19757  M20-2586-10  Modulex Box - 10x Window 1 x 8 x 6  1963       716   
19758  M20-2586-10  Modulex Box - 10x Window 1 x 8 x 6  1963       716   
19759  WEETABIX1-1                     Weetabix Castle  1970       414   

       num_parts  id_x name_themes  parent_id    id_y  version  
0             43     1     Technic        NaN   24696        1  
1              3     1     Technic        NaN   24701        1  
2            211     1     Technic        NaN   15686        1  
3            120     1     Technic        NaN   13420        1  
4             39     1     Technic        NaN   10042        1  
...          ...   ...         ...        ...     ...      ...  
19755         10   716     Modulex        NaN  109713        1  
19756         10   716     Modulex        NaN  109714        2  
19757         10   716     Modulex        NaN  109715        1  
19758         10   716     Modulex        NaN  109716        2  
19759        471   414      Castle      411.0    5574        1  

[19760 rows x 10 columns]

In [46]:
inventory_minifigs

inventory_id     fig_num  quantity
0                 3  fig-001549         1
1                 4  fig-000764         1
2                19  fig-000555         1
3                25  fig-000574         1
4                26  fig-000842         1
...             ...         ...       ...
18139        122344  fig-009900         1
18140        122344  fig-009908         2
18141        122344  fig-009954         1
18142        122344  fig-009955         1
18143        122344  fig-009956         1

[18144 rows x 3 columns]

In [47]:
mini2 = pd.merge(mini1,inventory_minifigs, left_on = 'id_y', right_on='inventory_id')
mini2

set_num                         name_sets  year  theme_id  num_parts  \
0      8222-1                             V-TOL  1997         1        141   
1      8223-1  Hydrofoil 7 / Powerboat Columbia  1996         1        126   
2      8225-1         Road Rally V / Super Kart  1995         1         97   
3      8225-1         Road Rally V / Super Kart  1995         1         97   
4      8229-1                     Tread Trekker  1997         1        186   
...       ...                               ...   ...       ...        ...   
18139  9725-2        ROBOLAB Amusement Park Set  2002       520        270   
18140  9725-2        ROBOLAB Amusement Park Set  2002       520        270   
18141  9725-2        ROBOLAB Amusement Park Set  2002       520        270   
18142  9725-2        ROBOLAB Amusement Park Set  2002       520        270   
18143  9780-1      ROBOLAB Starter Building Set  1998       520       1712   

       id_x name_themes  parent_id    id_y  version  inventory_id     fig_num  \
0         1     Technic        NaN    9054        1          9054  fig-010502   
1         1     Technic        NaN     136        1           136  fig-010500   
2         1     Technic        NaN   10022        1         10022  fig-010488   
3         1     Technic        NaN   90471        2         90471  fig-010488   
4         1     Technic        NaN    6351        1          6351  fig-010488   
...     ...         ...        ...     ...      ...           ...         ...   
18139   520         RCX      518.0  103705        1        103705  fig-002229   
18140   520         RCX      518.0  103705        1        103705  fig-008698   
18141   520         RCX      518.0  103705        1        103705  fig-011555   
18142   520         RCX      518.0  103705        1        103705  fig-011556   
18143   520         RCX      518.0   10940        1         10940  fig-000467   

       quantity  
0             1  
1             1  
2             1  
3             1  
4             1  
...         ...  
18139         1  
18140         1  
18141         1  
18142         1  
18143         1  

[18144 rows x 13 columns]

In [48]:
minifigs

fig_num                                   name  num_parts
0      fig-000001                     Toy Store Employee          4
1      fig-000002                           Customer Kid          4
2      fig-000003                  Assassin Droid, White          8
3      fig-000004                           Basic Figure          4
4      fig-000005        Captain America with Short Legs          3
...           ...                                    ...        ...
11822  fig-012221                                Bob-omb         11
11823  fig-012222                   Boomer (Banzai) Bill         10
11824  fig-012223                            Bullet Bill         10
11825  fig-012224                           Buzzy Beetle          7
11826  fig-012225  Unikitty - Cheerykitty (Cheery Kitty)         14

[11827 rows x 3 columns]

In [49]:
mini3 = mini2.merge(minifigs, on='fig_num')
mini3

set_num                         name_sets  year  theme_id  num_parts_x  \
0      8222-1                             V-TOL  1997         1          141   
1      8223-1  Hydrofoil 7 / Powerboat Columbia  1996         1          126   
2      8714-1               TECHNIC Team / Team  1993        19            9   
3      8225-1         Road Rally V / Super Kart  1995         1           97   
4      8225-1         Road Rally V / Super Kart  1995         1           97   
...       ...                               ...   ...       ...          ...   
18139  9355-1                         Space Set  1990       527          559   
18140  9355-1                         Space Set  1990       527          559   
18141  9376-1                        Castle Set  1997       512          337   
18142  9376-1                        Castle Set  1997       512          337   
18143  9725-2        ROBOLAB Amusement Park Set  2002       520          270   

       id_x   name_themes  parent_id    id_y  version  inventory_id  \
0         1       Technic        NaN    9054        1          9054   
1         1       Technic        NaN     136        1           136   
2        19  Supplemental        1.0    4932        1          4932   
3         1       Technic        NaN   10022        1         10022   
4         1       Technic        NaN   90471        2         90471   
...     ...           ...        ...     ...      ...           ...   
18139   527         Space      507.0    3140        1          3140   
18140   527         Space      507.0    3140        1          3140   
18141   512        Castle      507.0    4857        1          4857   
18142   512        Castle      507.0    4857        1          4857   
18143   520           RCX      518.0  103705        1        103705   

          fig_num  quantity  \
0      fig-010502         1   
1      fig-010500         1   
2      fig-010500         1   
3      fig-010488         1   
4      fig-010488         1   
...           ...       ...   
18139  fig-006948         4   
18140  fig-006949         4   
18141  fig-006552         2   
18142  fig-007104         2   
18143  fig-011556         1   

                                                    name  num_parts_y  
0      Technic Figure Blue Legs, White Top with Zip &...            3  
1      Technic Figure White Legs, White Top with Red ...            3  
2      Technic Figure White Legs, White Top with Red ...            3  
3      Technic Figure Black Legs, Light Gray Top with...            3  
4      Technic Figure Black Legs, Light Gray Top with...            3  
...                                                  ...          ...  
18139  Futuron Spaceman, Black and Yellow with Airtan...            6  
18140  Futuron Spaceman, Red and Green with Airtanks ...            6  
18141  Knight (Royal Knights) - Black Legs, Dark Gray...            5  
18142       King (Royal Knights) - Blue Legs, Black Hips            4  
18143     Woman, Blue Torso, Black Legs, Dark Flesh Hair            4  

[18144 rows x 15 columns]

In [50]:
Q13 = int(mini3[mini3['name_sets']== 'Millennium Falcon']['name'].nunique()/4)
Q13


8

**Q14:** 
- Create a function called `get_minifigs` that takes as input a set number (as a string) and a version number (as an integer) and outputs all minifigs from that set / version combination.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `set_num`, `fig_num`, `fig_name`, `fig_num_parts`.  
- The index should go from 0 to n-1.

In [51]:
mini3.head()

set_num                         name_sets  year  theme_id  num_parts_x  \
0  8222-1                             V-TOL  1997         1          141   
1  8223-1  Hydrofoil 7 / Powerboat Columbia  1996         1          126   
2  8714-1               TECHNIC Team / Team  1993        19            9   
3  8225-1         Road Rally V / Super Kart  1995         1           97   
4  8225-1         Road Rally V / Super Kart  1995         1           97   

   id_x   name_themes  parent_id   id_y  version  inventory_id     fig_num  \
0     1       Technic        NaN   9054        1          9054  fig-010502   
1     1       Technic        NaN    136        1           136  fig-010500   
2    19  Supplemental        1.0   4932        1          4932  fig-010500   
3     1       Technic        NaN  10022        1         10022  fig-010488   
4     1       Technic        NaN  90471        2         90471  fig-010488   

   quantity                                               name  num_parts_y  
0         1  Technic Figure Blue Legs, White Top with Zip &...            3  
1         1  Technic Figure White Legs, White Top with Red ...            3  
2         1  Technic Figure White Legs, White Top with Red ...            3  
3         1  Technic Figure Black Legs, Light Gray Top with...            3  
4         1  Technic Figure Black Legs, Light Gray Top with...            3

In [52]:

mini3.rename(columns={'name':'fig_name', 'num_parts_y':'fig_num_parts'}, inplace=True)

def get_minifigs(set_num, version):
    return mini3[(mini3['set_num'] == set_num) & 
                 (mini3['version'] == version)][['set_num',
                                                 'fig_num', 'fig_name', 
                                                 'fig_num_parts']].sort_values('fig_num').reset_index(drop=True)




**Q15:** Using the above function, output information about the minifigs from the set discussed in the `Millennium Falcon Prep Exercise`.  Save this output as `Q15`.

In [53]:
Q15 = get_minifigs('75192-1',1)
Q15

set_num     fig_num                                           fig_name  \
0  75192-1  fig-001714                              Chewbacca, Dark Brown   
1  75192-1  fig-001810                             Finn, Black Undershirt   
2  75192-1  fig-002057                                 Rey, Dark Tan Robe   
3  75192-1  fig-002514    C-3PO, Pearl Gold, Colorful Wires, Printed Legs   
4  75192-1  fig-002544                                               BB-8   
5  75192-1  fig-002549  Princess Leia, Hoth Outfit, White Jacket, Prin...   
6  75192-1  fig-002550  Han Solo, Dark Blue Jacket, Dark Brown Legs, P...   
7  75192-1  fig-002551                               Han Solo, Old, Angry   

   fig_num_parts  
0              3  
1              4  
2              4  
3              3  
4              2  
5              4  
6              4  
7              4

**Code Check:** Make sure that your `get_minifigs()` function works with both version 1 and version 2 of set 10217-1, Harry Potter's Diagon Alley.  Both versions should have 11 minifigs in the DataFrame.

In [54]:
get_minifigs('10217-1',1)

set_num     fig_num                                           fig_name  \
0   10217-1  fig-002229                   Skeleton (Standard) (3626b Head)   
1   10217-1  fig-006075                 Rubeus Hagrid, Dark Brown Overcoat   
2   10217-1  fig-006084  Lucius Malfoy, Black Robes and Cape, Death Eat...   
3   10217-1  fig-006130  Harry Potter, Open Dark Blue Jacket over White...   
4   10217-1  fig-006138                                    Fenrir Greyback   
5   10217-1  fig-006162  Ron Weasley, Red Tartan Sweater Vest, Reddish ...   
6   10217-1  fig-006163          Fred / George Weasley in Dark Orange Suit   
7   10217-1  fig-006164                                 Garrick Ollivander   
8   10217-1  fig-006165    Goblin, Black Jacket and Light Bluish Gray Legs   
9   10217-1  fig-006166        Goblin, Black Jacket and Reddish Brown Legs   
10  10217-1  fig-006167  Hermione Granger, Light Bluish Gray Sweater wi...   

    fig_num_parts  
0               6  
1               3  
2               5  
3               4  
4               4  
5               4  
6               4  
7               4  
8               3  
9               3  
10              4

**Q16:** 
- Create a function called `minifigs_from_themes` that takes as input a theme id (as an integer) and outputs all the minifigs contained in sets with that theme.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `fig_num`, `fig_name`. 
- The index should go from 0 to n-1.  
- Exact duplicates should be removed so that only unique figs are included.

In [55]:
def minifigs_from_themes(theme_id):
    return mini3[mini3['theme_id'] == theme_id][['fig_num', 'fig_name']].drop_duplicates().sort_values('fig_num').reset_index(drop=True)
    


**Code Check:** Call the `minifigs_from_themes()` function using the theme number of the set discussed in the `Millennium Falcon Prep Exercise`.  There should be 149 minifigs output for that respective theme.

In [56]:
minifigs_from_themes(171)

fig_num                                           fig_name
0    fig-000003                              Assassin Droid, White
1    fig-000124                           Chewbacca, Reddish Brown
2    fig-000144  Luke Skywalker, White Robe, White Legs, Cheek ...
3    fig-000303           Clone Trooper, Phase I Armor, Brown Eyes
4    fig-000516  Darth Vader, Light Bluish Gray Skin, Scarred R...
..          ...                                                ...
144  fig-011942  AT-AT Driver - Light Bluish Grey Helmet, Sand ...
145  fig-011944  Snowtrooper, White Hips, Reddish Brown Head, F...
146  fig-011945  Snowtrooper, White Hips, Light Flesh Head, Female
147  fig-011946  Snowtrooper, White Hips, Light Flesh Head, Off...
148  fig-011957    Snowtrooper, White Hips, Light Flesh Head, Grin

[149 rows x 2 columns]

**Q17:** 
- Create a function called `sets_from_minifig` that takes as input a minifig number (as a string) and returns a list of sets a minifig has appeared in.  
- Return a DataFrame (sorted by `set_num` and then by `version`) with the following columns: `set_num`, `version`,`name_sets`, `fig_num`, `fig_name`.  
- The index should go from 0 to n-1.

In [57]:
def sets_from_minifig(fig_num):
    return mini3[mini3['fig_num'] == fig_num][['set_num', 
                                               'version', 'name_sets', 'fig_num', 
                                               'fig_name']].sort_values(['set_num', 'version']).reset_index(drop=True)


**Code Check:** Call your `sets_from_minifig()` function on one of the Lego Friends minifigs, Andrea, who has figure number: `fig-001039`.  Your function should return a DataFrame that contains the following sets:

- Andrea on the Beach
- Birthday Party
- City Park Cafe

In [58]:
sets_from_minifig('fig-001039')

set_num  version            name_sets     fig_num  \
0  30100-1        1  Andrea on the Beach  fig-001039   
1  30107-1        1       Birthday Party  fig-001039   
2   3061-1        1       City Park Café  fig-001039   

                                            fig_name  
0  Andrea - Bright Light Orange Top, Light Aqua S...  
1  Andrea - Bright Light Orange Top, Light Aqua S...  
2  Andrea - Bright Light Orange Top, Light Aqua S...

**Q18:** 

*Note:  Before you begin working on this question, please note that CodeGrade has a 1GB RAM limit.  If you see an `Exit -9` code when submitting your assignment, it is probably due to this RAM limit and previous students had an issue with this question causing too much memory usage.  The biggest thing to think about is to only use the tables that you need to answer for this question.  For instance, you do not need any of the minifigs data since this is not needed for the output and the merging of this data could cause the number of rows to "explode" in terms of the total count.  I understand that this can be annoying that we have to work around CodeGrade's limitations, but it is good practice in memory management.*

*You should only need to merge the following datasets for this answer so as to not go above the RAM limit: `themes`, `sets`, `inventories`, `inventory_parts`, `parts`, and `colors`*

- Create a function called `set_parts` that takes as input the set number (as a string) and version number (as an integer) and outputs a DataFrame with each part in the set.  
- The DataFrame should have the following columns/column names: `set_num`, `version`, `name_sets`, `year`, `name_themes`, `part_num`, `name_part`, `name_color`,`quantity`,`is_spare`.  
- The Data should be sorted by `part_num` and then `name_color`.  
- The index should go from 0 to n-1.

In [59]:
inventory_parts.head()

inventory_id        part_num  color_id  quantity is_spare
0             1        48379c01        72         1        f
1             1           48395         7         1        f
2             1  stickerupn0077      9999         1        f
3             1         upn0342         0         1        f
4             1         upn0350        25         1        f

In [60]:
inventories.head()

id  version set_num
0   1        1  7922-1
1   3        1  3931-1
2   4        1  6942-1
3  15        1  5158-1
4  16        1   903-1

In [61]:
mini1.head()

set_num                       name_sets  year  theme_id  num_parts  id_x  \
0   001-1                           Gears  1965         1         43     1   
1   002-1  4.5V Samsonite Gears Motor Set  1965         1          3     1   
2  1030-1  TECHNIC I: Simple Machines Set  1985         1        211     1   
3  1038-1              ERBIE the Robo-Car  1985         1        120     1   
4  1039-1            Manual Control Set 1  1986         1         39     1   

  name_themes  parent_id   id_y  version  
0     Technic        NaN  24696        1  
1     Technic        NaN  24701        1  
2     Technic        NaN  15686        1  
3     Technic        NaN  13420        1  
4     Technic        NaN  10042        1

In [62]:
#themesandsets = pd.merge(sets,themes, left_on = 'theme_id', right_on='id')
#mini1 = themesandsets.merge(inventories, on='set_num')
add_inventory_parts = pd.merge(mini1,inventory_parts,left_on = 'id_y', right_on='inventory_id' )
add_inventory_parts
        

set_num        name_sets  year  theme_id  num_parts  id_x  \
0             001-1            Gears  1965         1         43     1   
1             001-1            Gears  1965         1         43     1   
2             001-1            Gears  1965         1         43     1   
3             001-1            Gears  1965         1         43     1   
4             001-1            Gears  1965         1         43     1   
...             ...              ...   ...       ...        ...   ...   
935303  WEETABIX1-1  Weetabix Castle  1970       414        471   414   
935304  WEETABIX1-1  Weetabix Castle  1970       414        471   414   
935305  WEETABIX1-1  Weetabix Castle  1970       414        471   414   
935306  WEETABIX1-1  Weetabix Castle  1970       414        471   414   
935307  WEETABIX1-1  Weetabix Castle  1970       414        471   414   

       name_themes  parent_id   id_y  version  inventory_id   part_num  \
0          Technic        NaN  24696        1         24696       132a   
1          Technic        NaN  24696        1         24696       3020   
2          Technic        NaN  24696        1         24696      3062c   
3          Technic        NaN  24696        1         24696   3404bc01   
4          Technic        NaN  24696        1         24696         36   
...            ...        ...    ...      ...           ...        ...   
935303      Castle      411.0   5574        1          5574      3062a   
935304      Castle      411.0   5574        1          5574      3062a   
935305      Castle      411.0   5574        1          5574   3081cc01   
935306      Castle      411.0   5574        1          5574      700ed   
935307      Castle      411.0   5574        1          5574  777pr0018   

        color_id  quantity is_spare  
0              7         4        f  
1             15         4        f  
2             15         1        f  
3             15         4        f  
4              7         4        f  
...          ...       ...      ...  
935303         0         8        f  
935304        15         2        f  
935305        15         2        f  
935306         7         3        f  
935307        15         1        f  

[935308 rows x 15 columns]

In [63]:
add_parts = pd.merge(add_inventory_parts,parts,on = 'part_num')
add_parts.head()

set_num            name_sets  year  theme_id  num_parts  id_x name_themes  \
0   001-1                Gears  1965         1         43     1     Technic   
1   003-1  Master Mechanic Set  1966       366        403   366   Basic Set   
2   021-1            Wheel Set  1966       366        183   366   Basic Set   
3   033-2   Basic Building Set  1968       366        177   366   Basic Set   
4   044-1   Basic Building Set  1968       366        225   366   Basic Set   

   parent_id   id_y  version  inventory_id part_num  color_id  quantity  \
0        NaN  24696        1         24696     132a         7         4   
1      365.0  24702        1         24702     132a         7         8   
2      365.0  24734        1         24734     132a         7         6   
3      365.0   3372        1          3372     132a         0         4   
4      365.0   6772        1          6772     132a         0         4   

  is_spare                           name  part_cat_id part_material  
0        f  Tyre Smooth Old Style - Small           29        Rubber  
1        f  Tyre Smooth Old Style - Small           29        Rubber  
2        f  Tyre Smooth Old Style - Small           29        Rubber  
3        f  Tyre Smooth Old Style - Small           29        Rubber  
4        f  Tyre Smooth Old Style - Small           29        Rubber

In [64]:
all_data = pd.merge(add_parts,colors,left_on = 'color_id', right_on='id' )
#all_data.head()
all_data.rename(columns={'name_x':'name_part','name_y':'name_color'}, inplace=True)
all_data.head()

set_num                      name_sets  year  theme_id  num_parts  id_x  \
0   001-1                          Gears  1965         1         43     1   
1   003-1            Master Mechanic Set  1966       366        403   366   
2   021-1                      Wheel Set  1966       366        183   366   
3   080-1  Basic Building Set with Train  1967       366        712   366   
4   223-1                  Adventure Set  1968       366        223   366   

  name_themes  parent_id   id_y  version  inventory_id part_num  color_id  \
0     Technic        NaN  24696        1         24696     132a         7   
1   Basic Set      365.0  24702        1         24702     132a         7   
2   Basic Set      365.0  24734        1         24734     132a         7   
3   Basic Set      365.0   1973        1          1973     132a         7   
4   Basic Set      365.0  25381        1         25381     132a         7   

   quantity is_spare                      name_part  part_cat_id  \
0         4        f  Tyre Smooth Old Style - Small           29   
1         8        f  Tyre Smooth Old Style - Small           29   
2         6        f  Tyre Smooth Old Style - Small           29   
3         8        f  Tyre Smooth Old Style - Small           29   
4         4        f  Tyre Smooth Old Style - Small           29   

  part_material  id  name_color     rgb is_trans  
0        Rubber   7  Light Gray  9BA19D        f  
1        Rubber   7  Light Gray  9BA19D        f  
2        Rubber   7  Light Gray  9BA19D        f  
3        Rubber   7  Light Gray  9BA19D        f  
4        Rubber   7  Light Gray  9BA19D        f

In [65]:
def set_parts(set_num, version):
    set_version= all_data[(all_data['set_num'] == set_num) & (all_data['version'] == version)]
    part_col = ['set_num', 'version', 'name_sets', 'year', 'name_themes', 'part_num', 'name_part', 'name_color', 'quantity', 'is_spare']
    return set_version[part_col].sort_values(['part_num', 'name_color']).reset_index(drop=True)
    


**Code Check**: Using the `set_parts()` function, output a list of parts for the set discussed in the `Millennium Falcon Prep Exercise`. There should be a total of 730 rows in the DataFrame, which includes 676 parts and 54 spare parts. You can check the respective page on the rebrickable website to verify these results.

In [66]:
set_parts('75192-1',1)

set_num  version          name_sets  year                name_themes  \
0    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
1    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
2    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
3    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
4    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
..       ...      ...                ...   ...                        ...   
725  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
726  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
727  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
728  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
729  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   

    part_num                                          name_part  \
0      10197  Technic Pin Connector Hub with 2 Perpendicular...   
1      10201            Bracket 1 x 2 - 1 x 4 [Rounded Corners]   
2      10247  Plate Special 2 x 2 with 1 Pin Hole [Complete ...   
3      10247  Plate Special 2 x 2 with 1 Pin Hole [Complete ...   
4      10314  Brick Curved 1 x 4 x 1 1/3 No Studs, Curved To...   
..       ...                                                ...   
725    99563  Tile Special 1 x 2 with Sloped Walls AKA Money...   
726    99780                     Bracket 1 x 2 - 1 x 2 Inverted   
727    99780                     Bracket 1 x 2 - 1 x 2 Inverted   
728    99781                              Bracket 1 x 2 - 1 x 2   
729    99784  Bar 1 x 12 with 1 x 2 Plate End with Hollow St...   

            name_color  quantity is_spare  
0     Dark Bluish Gray         3        f  
1    Light Bluish Gray        12        f  
2     Dark Bluish Gray         5        f  
3                White        22        f  
4                  Tan         3        f  
..                 ...       ...      ...  
725  Light Bluish Gray        70        f  
726              Black         6        f  
727   Dark Bluish Gray        25        f  
728  Light Bluish Gray         2        f  
729  Light Bluish Gray         3        f  

[730 rows x 10 columns]

In [67]:
set_parts('75192-1',1)[set_parts('75192-1',1)['is_spare'] == 't'].count()

set_num        54
version        54
name_sets      54
year           54
name_themes    54
part_num       54
name_part      54
name_color     54
quantity       54
is_spare       54
dtype: int64

In [68]:
set_parts('75192-1',1).shape[0] - 54

676